llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -8379,7 +8383,7 @@ static bool matchRotateSub(SDValue Pos, SDValue Neg, unsigned EltSize,
   // NOTE: We can only do this when matching operations which won't modify the
   // least Log2(EltSize) significant bits and not a general funnel shift.
   unsigned MaskLoBits = 0;
-  if (IsRotate && isPowerOf2_64(EltSize)) {
+  if (IsRotate && !FromAdd && isPowerOf2_64(EltSize)) {
```

Comment:
Is it incorrect to do this for `add`? Or do we just not want to?

Reply:
It is incorrect to do this for an ADD or a non-rotate funnel-shift because when the shift amount is 0 the LHS and RHS will be combined with each other. If they are equal (IsRotate) and the combining operation is an OR (!FromAdd) than this will be a no-op (LHS = RHS = Result) which is consistent with a funnel-shift of 0. If either of these conditions are not met the then they cannot be safely combined like this for the 0 case.